ducc/lib/conversion.go

```diff
@@ -32,27 +33,151 @@ const (
+	if majorI >= 4 {
+		return nil
+	}
+	minorS := sems[1]
+	minorI, err := strconv.Atoi(minorS)
+	if err != nil {
+		errF := fmt.Errorf("Singularity version returned an unexpected format, unable to parse Minor number: %s", err)
+		LogE(errF).WithFields(log.Fields{"version": version, "minor number": minorS}).Error("Not valid singularity")
+		return errF
+	}
+	if majorI >= 3 && minorI >= 5 {
+		return nil
+	}
+	errF := fmt.Errorf("Installed singularity is too old, we need at least 3.5: Installed version: %s", version)
+	LogE(errF).WithFields(log.Fields{"version": version}).Error("Too old singularity")
+	return errF
+}
+
 func ConvertWishSingularity(wish WishFriendly) (err error) {
+	err = assureValidSingularity()
+	if err != nil {
+		return err
+	}
 	tmpDir, err := UserDefinedTempDir("", "conversion")
 	if err != nil {
 		LogE(err).Error("Error when creating tmp singularity directory")
 		return
 	}
 	defer os.RemoveAll(tmpDir)
 	inputImage, err := ParseImage(wish.InputName)
-	singularity, err := inputImage.DownloadSingularityDirectory(tmpDir)
-	defer os.RemoveAll(singularity.TempDirectory)
+	expandedImgTag, err := inputImage.ExpandWildcard()
 	if err != nil {
-		LogE(err).Error("Error in dowloading the singularity image")
+		LogE(err).WithFields(log.Fields{
+			"input image": fmt.Sprintf("%s/%s", inputImage.Registry, inputImage.Repository)}).
+			Error("Error in retrieving all the tags from the image")
 		return
 	}
+	var firstError error
+	for _, inputImage := range expandedImgTag {
+		// we want to check if we have already ingested the Singularity image
+		// Several cases are possible
+		// Image not ingested, neither pubSymPath nor privatePath are present
+		// Image ingested and up to date, pubSymPath and privatePath point to the same thing
+		// Image update but stale (old), pubSymPath and privatePath point to different things
+		publicSymlinkPath := inputImage.GetPublicSymlinkPath()
+		completePubSymPath := filepath.Join("/", "cvmfs", wish.CvmfsRepo, publicSymlinkPath)
+		_, errPub := os.Stat(completePubSymPath)
+
+		singularityPrivatePath, err := inputImage.GetSingularityPath()
+		if err != nil {
+			errF := fmt.Errorf("Error in getting the path where to save Singularity filesystem: %s", err)
+			LogE(err).Warning(errF)
+			firstError = errF
+			continue
+		}
+		completeSingularityPriPath := filepath.Join("/", "cvmfs", wish.CvmfsRepo, singularityPrivatePath)
+		_, errPri := os.Stat(completeSingularityPriPath)
 
-	err = singularity.IngestIntoCVMFS(wish.CvmfsRepo)
-	if err != nil {
-		LogE(err).Error("Error in ingesting the singularity image into the CVMFS repository")
-		return
+		// no error in stating both directories
+		// either the image is up to date or the image became stale
+		if errPub == nil && errPri == nil {
+			pubLinkPointsTo, err := os.Readlink(completePubSymPath)
+			if err != nil {
+				// this should never happen
+				Log().Warning("Error in reading the public link")
+			}
+			if pubLinkPointsTo == singularityPrivatePath {
```

Comment:
This check is wrong!

We should check if they point to the same directory, not if the path is the same.
Indeed one is a relative path and the other is a "direct" path.

We can stat it, or we already do, and compare the inode :)

Reply:
The equivalent of bash's `readlink --canonicalize`?

Reply:
Exactly!